helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -27,21 +27,21 @@
  * Evaluate a partition allocation proposal and return YES or NO based on the cluster context.
  * Any proposal fails one or more hard constraints will be rejected.
  */
-abstract class HardConstraint {
+public abstract class HardConstraint {
```

Comment:
I'm fine if we need a public one. But where we need it? It is not in the interface, right?

Reply:
It's needed because OptimalAssignment is a different package than the constraint. I think public interface is fine as long as the detail implementation is hidden.

Reply:
What you really need is the description, right?
Actually, I do have a concern if you refer to the hard constraint in the optimal assignment. Note that the constraint algorithm is just one of the option. But the optimal assignment shall be applicable to all different algorithms. If you are directly referring to the constraint, how you are going to support the other algorithm which does not know about constraint?

Reply:
@jiajunwang You mean to say, OptimalAssignment shouldn't know about constraints? As in OptimalAssignment only contains the final assignment information, and that assignment isn't necessary tied to ConstraintRebalancer, and for example, could also be used to store results of, say, RandomRebalancer, or NoMovementRebalancer (I do see some applications that might benefit from NoMovement...)